From 19df977cae7ce8d6708ab58dd7271c259b296ad2 Mon Sep 17 00:00:00 2001
From: RealJohnGalt <johngalt@fake.mail>
Date: Thu, 6 Jul 2017 23:30:58 -0700
Subject: [PATCH] AMLUtils: support Nougat-kernel fractional rates

Initially implemented by @kszaq for Krypton, this is now updated for Leia.

This also always sets mode to pass colorspace changes. Previously there
was an issue for instance if GUI was bt.709@50hz and bt.2020@25fps
played, the mode wouldn't be set and the colorspace would stay bt.709.
---
 xbmc/utils/AMLUtils.cpp | 45 ++++++++++++++++++++++++++++++++++-----------
 xbmc/utils/AMLUtils.h   |  2 +-
 2 files changed, 35 insertions(+), 12 deletions(-)

diff --git a/xbmc/utils/AMLUtils.cpp b/xbmc/utils/AMLUtils.cpp
index c674655..68aa97a 100644
--- a/xbmc/utils/AMLUtils.cpp
+++ b/xbmc/utils/AMLUtils.cpp
@@ -585,18 +585,22 @@ bool aml_get_native_resolution(RESOLUTION_INFO *res)
 {
   std::string mode;
   SysfsUtils::GetString("/sys/class/display/mode", mode);
-  return aml_mode_to_resolution(mode.c_str(), res);
+  bool result = aml_mode_to_resolution(mode.c_str(), res);
+
+  int fractional_rate;
+  SysfsUtils::GetInt("/sys/class/amhdmitx/amhdmitx0/frac_rate_policy", fractional_rate);
+  if (fractional_rate == 1)
+    res->fRefreshRate /= 1.001;
+
+  return result;
 }
 
 bool aml_set_native_resolution(const RESOLUTION_INFO &res, std::string framebuffer_name)
 {
   bool result = false;
 
-  // Don't set the same mode as current
-  std::string mode;
-  SysfsUtils::GetString("/sys/class/display/mode", mode);
-  if (res.strId != mode)
-    result = aml_set_display_resolution(res.strId.c_str(), framebuffer_name);
+  // Always set mode to pass colorspace changes
+  result = aml_set_display_resolution(res, framebuffer_name);
 
   aml_handle_scale(res);
 
@@ -626,6 +629,18 @@ bool aml_probe_resolutions(std::vector<RESOLUTION_INFO> &resolutions)
     {
       if (aml_mode_to_resolution(i->c_str(), &res))
         resolutions.push_back(res);
+
+      switch ((int)res.fRefreshRate)
+      {
+        case 24:
+        case 30:
+        case 60:
+          res.fRefreshRate /= 1.001;
+          res.strMode       = StringUtils::Format("%dx%d @ %.2f%s - Full Screen", res.iScreenWidth, res.iScreenHeight, res.fRefreshRate,
+            res.dwFlags & D3DPRESENTFLAG_INTERLACED ? "i" : "");
+          resolutions.push_back(res);
+          break;
+      }
     }
   }
   return resolutions.size() > 0;
@@ -641,13 +656,21 @@ bool aml_get_preferred_resolution(RESOLUTION_INFO *res)
   return true;
 }
 
-bool aml_set_display_resolution(const char *resolution, std::string framebuffer_name)
+bool aml_set_display_resolution(const RESOLUTION_INFO &res, std::string framebuffer_name)
 {
-  std::string mode = resolution;
+  std::string mode = res.strId.c_str();
+  std::string cur_mode;
+
   // switch display resolution
+  SysfsUtils::GetString("/sys/class/display/mode", cur_mode);
+
+  if (cur_mode == mode)
+    SysfsUtils::SetString("/sys/class/display/mode", "null");
+
+  int fractional_rate = (res.fRefreshRate == floor(res.fRefreshRate)) ? 0 : 1;
+  SysfsUtils::SetInt("/sys/class/amhdmitx/amhdmitx0/frac_rate_policy", fractional_rate);
+
   SysfsUtils::SetString("/sys/class/display/mode", mode.c_str());
 
-  RESOLUTION_INFO res;
-  aml_mode_to_resolution(mode.c_str(), &res);
   aml_set_framebuffer_resolution(res, framebuffer_name);
 
diff --git a/xbmc/utils/AMLUtils.h b/xbmc/utils/AMLUtils.h
index 5323ae9..df2f6d6 100644
--- a/xbmc/utils/AMLUtils.h
+++ b/xbmc/utils/AMLUtils.h
@@ -67,7 +67,7 @@ bool aml_get_native_resolution(RESOLUTION_INFO *res);
 bool aml_set_native_resolution(const RESOLUTION_INFO &res, std::string framebuffer_name);
 bool aml_probe_resolutions(std::vector<RESOLUTION_INFO> &resolutions);
 bool aml_get_preferred_resolution(RESOLUTION_INFO *res);
-bool aml_set_display_resolution(const char *resolution, std::string framebuffer_name);
+bool aml_set_display_resolution(const RESOLUTION_INFO &res, std::string framebuffer_name);
 void aml_setup_video_scaling(const char *mode);
 void aml_handle_scale(const RESOLUTION_INFO &res);
 void aml_enable_freeScale(const RESOLUTION_INFO &res);
-- 
2.13.0

